### PR TITLE
Patch @typecheck() to support containers as outputs

### PR DESCRIPTION
**Pull request by titu1994**
# Changelog

- `@typecheck()` will now enforce that `input_types()` and `output_types()` are defined, not `input_ports()` and `output_ports()`.
- `@typecheck()` will now recursively attempt to attach neural type to outputs which are of type list / tuple. Note- it will fail silently if types cannot be attached (say items of container are primitive types)
- `@typecheck()` will now test number of output ports and number of provided outputs match exactly.
- Add unittests for `@typecheck()`

Signed-off-by: smajumdar <titu1994@gmail.com>

### CODE REVIEW

File: nemo/core/classes/common.py [hunk 1]
```diff
@@ -74,11 +75,39 @@ def _attach_and_validate_output_types(self, out_objects):
         # TODO: Properly implement this
```
Comment (blisc) [inline nemo/core/classes/common.py:75]: To double-check that I understand this function, would the following docstring make sense?
`
This function does two things. 1) It ensures that len(out_object) == len(self.output_types). 2) If the output is a tensor (or list/tuple of list/tuple ... of tensors), it attaches a neural_type to it. Note: tensor.neural_type is only checked on during _validate_input_types which is called prior to forward().
`

File: nemo/core/classes/common.py [hunk 2]
```diff
@@ -74,11 +75,39 @@ def _attach_and_validate_output_types(self, out_objects):
...
+                        len(out_container), len(self.output_types)
+                    )
+                )
+
+            # Attach types recursively, if possible
             if not isinstance(out_objects, tuple) and not isinstance(out_objects, list):
-                out_objects.neural_type = out_types_list[0][1]
+                try:
+                    out_objects.neural_type = out_types_list[0][1]
+                except Exception:
```
Comment (blisc) [inline nemo/core/classes/common.py:96]: Let's not catch Exception, can we either find out exactly what error is raised and catch that? Alternatively, can we somehow define our own exception?

File: nemo/core/classes/common.py [hunk 3]
```diff
@@ -74,11 +75,39 @@ def _attach_and_validate_output_types(self, out_objects):
...
+                    self.__attach_neural_type(res, out_types_list[ind][1])
+
+    def __attach_neural_type(self, obj, type_val):
+        if isinstance(obj, tuple) or isinstance(obj, list):
+            for elem in obj:
+                self.__attach_neural_type(elem, type_val)
+
+        try:
+            obj.neural_type = type_val
+        except Exception:
```
Comment (blisc) [inline nemo/core/classes/common.py:109]: Let's not catch Exception, can we either find out exactly what error is raised and catch that? Alternatively, can we somehow define our own exception?

File: nemo/core/classes/common.py (hunk 2 above)
Comment (titu1994) [reply nemo/core/classes/common.py:96]: It depends on the type of object you submit, and how it reacts to an attempt to assign a object instance to it. List behaves differently than tuple, than base data types, than structs, than data classes. There's too many combinations to check explicitly.

Comment (titu1994) [reply nemo/core/classes/common.py:96]: As to our own exception, we first need to catch something in order to raise it with our own. These exceptions comes from python core itself, no chance to override or wrap

File: nemo/core/classes/common.py (hunk 1 above)
Comment (titu1994) [reply nemo/core/classes/common.py:75]: This is a good start. Will add it with some clarifications. Technically we can attach types to anything which is an object of a class, not just tensors.

File: nemo/core/classes/common.py [hunk 4]
```diff
@@ -45,12 +45,18 @@ def output_types(self) -> Optional[Dict[str, NeuralType]]:
     def _validate_input_types(self, **kwargs):
         # TODO: Properly implement this
         if self.input_types is not None:
-            if len(kwargs) != len(self.input_types):
-                raise TypeError(
-                    "Number of input arguments provided ({}) is not as expected ({})".format(
-                        len(kwargs), len(self.input_types)
+            total_input_types = len(self.input_types)
+            mandatory_input_types = len(
+                [type_val for type_key, type_val in self.input_types.items() if not type_val.optional]
```
Comment (blisc) [inline nemo/core/classes/common.py:50]: How does one define that it is optional?

Comment (titu1994) [reply nemo/core/classes/common.py:50]: NeuralType has the argument optional = False by default

**Comment by tkornuta-nvidia** [comment]
> will now recursively attempt to attach neural type to outputs which are of type list / tuple. Note- it will fail silently if types cannot be attached (say items of container are primitive types)

Not sure if I understand it correctly... 

Example: Assume you have list of list of tuples (like: batch of lists of bounding boxes)
To which objects the type will be attached?

**Comment by titu1994** [comment]
Tuples / list themselves do now support assignments of attributes. Via recursion each individual element of the tuple will be tried and assigned the corresponding type, inside each list of lists.

If the list of list if tuples contains only primitive types, then even that assignment will fail and basically the entire recursion will not attach any type information

**Comment by tkornuta-nvidia** [comment]
> Tuples / list themselves do now support assignments of attributes. Via recursion each individual element of the tuple will be tried and assigned the corresponding type, inside each list of lists.

I see. And how do you define such a port?

**Comment by tkornuta-nvidia** [comment]
>> Example: Assume you have list of list of tuples (like: batch of lists of bounding boxes)

> If the list of list if tuples contains only primitive types, then even that assignment will fail and basically the entire recursion will not attach any type information

I do not understand... does it mean that the "real output" (list of list of 1-D tensor with four elements) will or won't have the type attached?

**Comment by titu1994** [comment]
As long as the individual elements of whatever list-type nest is an object of a class which supports argument assigment, they will individually assigned types. So the 1d-tensors themselves will have the `neural_type` value. However the list which contains these tensors will not have the `neural_type` - lists are containers, they dont support value assignment.

**Comment by titu1994** [comment]
> I see. And how do you define such a port?

I'm not particularly sure. As long as the container is homogeneous, then you can simply use the NeuralType of the homogeneous element itself and it will be fine.

File: nemo/core/classes/common.py (hunk 1 above)
Comment (titu1994) [reply nemo/core/classes/common.py:95]: Done

**Comment by tkornuta-nvidia** [comment]
> > I see. And how do you define such a port?
> 
> I'm not particularly sure. As long as the container is homogeneous, then you can simply use the NeuralType of the homogeneous element itself and it will be fine. For non-homogeneous containers, I'll defer to @okuchaiev

Ech, then I am not actually sure what this PR is about... I do not understand what will be happening with the axes in case of lists. Nor where types will be attached. Nor how they will be checked.

To make it clear: I am _just_ concerned, that's all.

**Comment by blisc** [comment]
> > > I see. And how do you define such a port?
> > 
> > 
> > I'm not particularly sure. As long as the container is homogeneous, then you can simply use the NeuralType of the homogeneous element itself and it will be fine. For non-homogeneous containers, I'll defer to @okuchaiev
> 
> Ech, then I am not actually sure what this PR is about... I do not understand what will be happening with the axes in case of lists. Nor where types will be attached. Nor how they will be checked.
> 
> To make it clear: I do not want to stop your work, I am _just_ concerned, that's all.

I am blocked because I need to return lists of tensors and the current typecheck does not support this.

File: nemo/collections/asr/models/classification_models.py [hunk 5]
```diff
@@ -40,7 +40,7 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         self.preprocessor = EncDecClassificationModel.from_config_dict(self._cfg.preprocessor)
         self.encoder = EncDecClassificationModel.from_config_dict(self._cfg.encoder)
         self.decoder = EncDecClassificationModel.from_config_dict(self._cfg.decoder)
-        self.loss = CrossEntropyLoss(logits_ndim=self.decoder.num_classes)
+        self.loss = CrossEntropyLoss()
```
Comment (blisc) [inline nemo/collections/asr/models/classification_models.py:43]: So how come you removed this from asr models but had to add it for the nlp models?

File: nemo/core/classes/common.py [hunk 6]
```diff
@@ -44,12 +44,18 @@ def output_types(self) -> Optional[Dict[str, NeuralType]]:
...
+            )
+
+            if len(kwargs) != total_input_types:
+                if len(kwargs) != mandatory_input_types:
+                    raise TypeError(
+                        "Number of input arguments provided ({}) is not as expected ({})".format(
+                            len(kwargs), len(self.input_types)
+                        )
```
Comment (blisc) [inline nemo/core/classes/common.py:57]: This if can be condensed

File: nemo/collections/asr/models/classification_models.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/models/classification_models.py:43]: The CrossEntropyLoss has the default value of `logits_ndim=2`. basically a flattened matrix shape of [batch, dimension] is mapped to this loss. This is the speech commands file, not asr, so that is the appropriate logits shape.

For NLP, they perform per timestep token generation, so logits are of shape [batch, timestep, dimension], and therefore I need to explicitly override the `logits_ndim=3` part for NLP. This has been verified by NLP team.

File: nemo/core/classes/common.py (hunk 6 above)
Comment (titu1994) [reply nemo/core/classes/common.py:74]: Good point. Done (in next commit)

**Comment by blisc** [comment]
Additional comments:
1) This functionality desperately needs unit tests
2) Why is output_types a dictionary if we don't use the dictionary keys? We should just make it a list

File: nemo/core/classes/common.py (hunk 6 above)
Comment (titu1994) [reply nemo/core/classes/common.py:74]: Spoke too soon. Cant condense it cause we need to disparate checks - 1 in case total_input_types match, and other if mandatory types match. If we use `and` or `or` we will always fail one of the two tests.

File: nemo/collections/asr/models/classification_models.py (hunk 5 above)
Comment (okuchaiev) [reply nemo/collections/asr/models/classification_models.py:43]: so ndim is shape rank, not number of classes, right?

**Comment by titu1994** [comment]
@blisc Agreed on tests, will add them to this PR
For the second point, historically Nemo never enforces the output type name to match (since these are tensors generated by pytorch). We still use dicts to define output_types since it is easier to construct and reason about - at one look we know index and order. Eventually, @okuchaiev can possibly enforce output types, but I still prefer both input and output_types to provide dictionary.

File: nemo/collections/nlp/models/token_classification/ner_model.py [hunk 7]
```diff
@@ -85,9 +85,9 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
 
         if cfg.class_balancing == 'weighted_loss':
             # You may need to increase the number of epochs for convergence when using weighted_loss
-            self.loss = CrossEntropyLoss(weight=self.data_desc.class_weights)
+            self.loss = CrossEntropyLoss(logits_ndim=3, weight=self.data_desc.class_weights)
```
Comment (okuchaiev) [inline nemo/collections/nlp/models/token_classification/ner_model.py:88]: @ekmb is this change OK?

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py [hunk 8]
```diff
@@ -89,7 +89,7 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
             use_transformer_init=cfg.capit_head.use_transformer_init,
         )
 
-        self.loss = CrossEntropyLoss()
+        self.loss = CrossEntropyLoss(logits_ndim=3)
```
Comment (okuchaiev) [inline nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py:92]: @ekmb  please have a look

File: nemo/collections/nlp/models/token_classification/ner_model.py (hunk 7 above)
Comment (titu1994) [reply nemo/collections/nlp/models/token_classification/ner_model.py:88]: I asked her on slack and she said its corect. Same for below.

**Comment by titu1994** [comment]
Lets wait to merge, im pushing tests to cover a good chunk of cases

**Comment by tkornuta-nvidia** [comment]
> For the second point, historically Nemo never enforces the output type name to match (since these are tensors generated by pytorch).

I have started to address that in Neural Graphs:
https://github.com/NVIDIA/NeMo/blob/b8619b6d12b14fcd77bb42e7d7ca84adc6b4673e/nemo/core/neural_graph.py#L1261
Actually get_batch() has two options: return a dict vs return a tuple.

Pointing as maybe you will find it useful.

File: nemo/collections/asr/models/classification_models.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/models/classification_models.py:43]: Yup thats right. I misunderstood the ndim part when I first wrote this line.

**Comment by titu1994** [comment]
@blisc I've added some (imo) pretty rigorous tests to cover many edge cases. Please see the last 2 test cases - `test_multi_forward_type` and `test_nested_input_output_neural_types`. @tkornuta-nvidia The nested test should give some idea of the rudimentary support for typing in nests for the moment. Its not the best, but its better than before (I think). And @okuchaiev can improve them by building on top of this later.

**Comment by titu1994** [comment]
@tkornuta-nvidia 
> Ech, then I am not actually sure what this PR is about... I do not understand what will be happening with the axes in case of lists. Nor where types will be attached. Nor how they will be checked.

PR is to unblock blisc, to fix the mismatch between `input/output_*ports*` and `input/output_*types*`, to add unittests for `@typecheck()` which was long overdue. 

For an example of how rudimentary nested typing works (in both homogeneous and heterogeneous cases of nesting), look at the second last test I added. This is not a global solution, but good enough to support nested list type outputs in the typing framework.

File: tests/core/test_typecheck.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+                _, u00 = w[0][0].max(-1)
+                _, u01 = w[0][1].max(-1)
+                _, u10 = w[1][0].max(-1)
+                return [[u00, u01], [u10]]
+
+        nodeA = NestedNodeA()
+        nodeB = NestedNodeB()
+
+        input_nest = [[torch.zeros(10), torch.zeros(10)], [torch.zeros(10)]]
+        outA = nodeA(x=input_nest)
```
Comment (tkornuta-nvidia) [inline tests/core/test_typecheck.py:318]: @titu1994 @okuchaiev I think this call should return an error: the input matching the definition: `NeuralType(('B',), ElementType()` should be `torch.zeros(10)` not `list of list of torch.zeros(10)`.

`List of list of torch.zeros(10)` should have this definition (incorrect syntax, just to show what I have in mind):
```
NeuralType(
  (Batch, is_list=True),
  (Time, is_list=True),
  (Depth, len=10),
   ElementType()
 )
```

This PR changes how type checking handles "dynamic axes" (i.e. axes that are lists). It seems that “dynamic” axes are skipped when type checking…? I really do not understand what is the logic of type checking now...
